lib/vector-core/src/event/vrl_target.rs

```diff
@@ -327,6 +327,137 @@ impl From<Event> for VrlTarget {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum VrlImmutableTarget<'a> {
+    LogEvent(&'a LogEvent),
+    Metric(&'a Metric),
+    Trace(&'a TraceEvent),
+}
+
+impl<'a> VrlImmutableTarget<'a> {
+    pub fn new(event: &'a Event) -> Self {
+        match event {
+            Event::Log(event) => VrlImmutableTarget::LogEvent(event),
+            Event::Metric(event) => VrlImmutableTarget::Metric(event),
+            Event::Trace(event) => VrlImmutableTarget::Trace(event),
+        }
+    }
+}
+
+impl<'a> vrl_lib::Target for VrlImmutableTarget<'a> {
+    fn target_insert(&mut self, _path: &LookupBuf, _value: vrl_lib::Value) -> Result<(), String> {
+        Err("cannot modify immutable target".to_string())
+    }
+
+    #[allow(clippy::redundant_closure_for_method_calls)] // false positive
+    fn target_get(&self, path: &LookupBuf) -> std::result::Result<Option<vrl_lib::Value>, String> {
```

Comment:
I'm assuming this has quite a bit of duplication? Any way we can reduce that?

Reply:
Yeah I can extract the metric code into a separate function, which I will do if we decide this change is worth keeping.